gtn/functions/compose.cpp

```diff
@@ -53,47 +86,81 @@ auto findReachable(
       reachable[idx] = true;
     }
     if (!epsilon_matched) {
-      for (auto i : first.in(curr.first)) {
-        if (first.olabel(i) != epsilon) {
-          if (first.olabelSorted()) {
-            // epsilon < 0
-            break;
-          } else {
-            continue;
-          }
-        }
-        auto un1 = first.srcNode(i);
-        auto idx = toIndex(un1, curr.second, first);
-        if (!reachable[idx]) {
-          // If we haven't seen this state before, explore it.
-          toExplore.emplace(un1, curr.second);
-        }
-        reachable[idx] = true;
-      }
-    }
-    if (!epsilon_matched) {
-      for (auto j : second.in(curr.second)) {
-        if (second.ilabel(j) != epsilon) {
-          if (second.ilabelSorted()) {
-            // epsilon < 0
-            break;
-          } else {
-            continue;
-          }
-        }
-        auto un2 = second.srcNode(j);
-        auto idx = toIndex(curr.first, un2, first);
-        if (!reachable[idx]) {
-          // If we haven't seen this state before, explore it.
-          toExplore.emplace(curr.first, un2);
-        }
-        reachable[idx] = true;
-      }
+      epsilonReachable(false, first, second, curr, reachable, toExplore);
+      epsilonReachable(true, first, second, curr, reachable, toExplore);
     }
   }
   return reachable;
 }
 
+bool addReachableNodeAndArc(
+    const Graph& first,
+    const Graph& second,
+    int currNode,
+    const std::pair<int, int> dstNodes,
```

Comment:
nit: could this also be a const ref to be consistent with below

Reply:
Yeah I can do that - I calculated a 64 bit reference is the same as 2 32-bit ints :)